### PR TITLE
[NNX] NNX migration prep (8/N): NNX native lora grpo

### PR DESCRIPTION
**Pull request by ecnal-cienet**
## NNX Migration Route Map

1. ✅ Add NNX scaffolding: `pure_nnx` flag, `init_state_fn`, `TrainStateNNX`, NNX utils. Linen workflow unchanged. (PR #3427)
2. ✅ NNX sharding utilities: `get_abstract_state_nnx`, `get_named_sharding_nnx`, `set_named_sharding_nnx`, `get_partition_spec_nnx`, `get_mesh_from_config`. (PR #3470)
3. ✅ NNX fully supported end-to-end: `TrainStateNNX`, model creation, gradient accumulation, checkpointing, and training loop dispatch. (PR #3500)
4. ✅ Sharding diagnostics on NNX, plus post-training bugfixes that surfaced once the NNX path got exercised end-to-end. (PR #3652)
4.5. ✅ Linen↔NNX checkpoint converter. (PR #3843)
4.6. ❌ Linen↔NNX checkpoint comparator (sibling branch on PR4.5).
5. ✅ NNX correctness fixes, feature enablements, and vocab tiling on NNX.
6. ✅ NNX-native DPO.
7. ✅ NNX-native MaxEngine inference. (PR #3821)
8. 🔄 **[This PR]** NNX-native LoRA + GRPO. NNX-native serving / decode-checkpoint LoRA via `apply_lora_on_base_params_nnx` / `unapply_lora_from_base_params_nnx` / `get_lora_abstract_state_nnx` (the maxengine `pure_nnx + LoRA` carve-out from PR7 is cleared); NNX-native GRPO trainer via `grpo_loss_fn_nnx` + `compute_log_probs_nnx` + NNX `setup_train_loop`/`train_step`/`eval_step` paths. Stacks on PR7.
9. ❌ NNX-aware QK-Clip + remaining checkpoint utilities.
9.5. ❌ NNX + AQT in MaxEngine + serve-mode reload + gpt3 prefill fix.
10. ❌ Vocab tiling `custom_vjp` for NNX.
11. ❌ Set NNX defaults to `True`; regenerate sharding goldens; flip back integration-test `pure_nnx=False` annotations.
12. ❌ Delete Linen-specific code paths and NNX compatibility flags.

## Description

This PR implements NNX-native LoRA serving and NNX-native GRPO by adding NNX-shape walkers and step helpers alongside the existing Linen ones, then dispatching on `config.pure_nnx`. Every NNX modification is gated by `if config.pure_nnx:`, preserving the Linen path byte-for-byte. The diff spans **+551 / −84 across 5 source files**, plus 2 new test files (515 lines).

## Part 1: NNX-shape LoRA Walkers

New helpers in `src/maxtext/utils/lora_utils.py` operating on `nnx.State` pure trees (no `{"params": ...}` outer wrap):

- `apply_lora_on_base_params_nnx` mutates `base_params` in place: `W += B @ A * scale` at target attention paths
- `unapply_lora_from_base_params_nnx` is the symmetric inverse
- `get_lora_abstract_state_nnx` walks the abstract `state.model` substate and emits a parallel tree with `lora_a.kernel`/`lora_b.kernel` leaves at target attention paths and `None` elsewhere
- `_nnx_param_subtree` drops the outer `TrainStateNNX` wrapping

The base model stays pristine; "apply" merges the delta into the kernel, "unapply" reverses. No `nnx.LoRA` wrapper, no model surgery. The on-disk format (HuggingFace PEFT-style `lora_a.kernel` / `lora_b.kernel`) round-trips between Linen and NNX consumers unchanged.

## Part 2: LoRA Dispatch in `setup_initial_lora_state` and `load_adapter`

Both top-level entry points in `lora_utils.py` branch on `config.pure_nnx`:
- NNX init builds the abstract base via `model_creation_utils.create_nnx_abstract_model` + `TrainStateNNX(model, optimizer)`
- Linen branch is the original `init_initial_state` + `get_lora_abstract_state` path, untouched

## Part 3: MaxEngine LoRA Carve-out Cleared

`src/maxtext/inference/maxengine/maxengine.py`:
- `load_single_adapter` no longer raises `NotImplementedError` on `pure_nnx`
- `apply_adapter` / `unapply_adapter` branch on `config.pure_nnx` to call the `_nnx` siblings

## Part 4: GRPO Loss and Step Helpers

`src/maxtext/experimental/rl/grpo_trainer.py`:

- `grpo_loss_fn_nnx(policy_model, config, data, dropout_rng, params, reference_model, is_train)`. Signature matches Linen `grpo_loss_fn` so callers dispatch on the same shape. `dropout_rng` and `params` are unused on NNX; `reference_model` is a frozen `nnx.Module` and the reference forward is wrapped in `stop_gradient`. Returns `(loss, LossAux)`, same dataclass as Linen.
- `_train_step_nnx`: `nnx.merge(graphdef, state)` to reconstruct `TrainStateNNX`, `value_and_grad` over policy params, `state.apply_gradients(grads)`, return `nnx.state(new_state, nnx.Not(nnx.Intermediate))`.
- `_eval_step_nnx`: same merge + loss-fn call, no state update.
- `train_step` / `eval_step` early-dispatch on `config.pure_nnx`; Linen branches verbatim.

## Part 5: GRPO setup_train_loop on NNX

`grpo_trainer.py::setup_train_loop`:
- Builds training and inference models via `mt.from_config(rngs=create_nnx_rngs(...))`
- Initializes state via `create_nnx_abstract_model` + `TrainStateNNX(model, optimizer, reference_model=...)`
- Reference uses the same init seed as policy and is never updated by `apply_gradients` (sibling field on `TrainStateNNX`, not embedded in `params`)
- The `WARNING: GRPO RL trainer does not yet support pure_nnx natively` log is removed

## Part 6: GRPO train_loop NNX Branches

`grpo_trainer.py::train_loop` — three Linen-coupled spots branched on `pure_nnx`:
- Initial reference seeding is skipped on NNX (already set up by `init_state_fn`)
- `metric_logger.write_setup_info_to_tensorboard` receives a flat `nnx.Param` state on NNX
- Checkpoint save passes the whole `TrainStateNNX` on NNX; the Linen `_split_grpo_state(state)[0]` strip is bypassed

The reshard call routes to `pathways_reshard_nnx` when `pure_nnx`. New helpers in `grpo_utils.py`:
- `compute_log_probs_nnx`: NNX model is called directly; intermediates pulled via `nnx.state(model, nnx.Intermediate).to_pure_dict()`
- `pathways_reshard_nnx`: splits `state.model` to a flat `nnx.Param` state, reshards onto the inference mesh, calls `inference_engine.update_params(...)`

## Part 7: Carve-outs (NotImplementedError Sites)

| Feature | Tracked In |
|---------|-----------|
| GRPO + `gradient_accumulation_steps > 1` | Follow-up |
| GRPO + `scan_layers=False` | Follow-up (needs an NNX-aware unscan helper) |

## Tests

**New unit tests** (`tests/unit/lora_utils_nnx_test.py`, 10 tests):
- 5 on `get_lora_abstract_state_nnx`: q/k/v/o shape derivation, target-vs-non-target masking, sharding propagation, leaf type validation, error paths
- 3 on `apply_lora_on_base_params_nnx`: apply→unapply identity, target-only mutation, numerical parity vs Linen `apply_lora_on_base_params` on the same random inputs
- 2 Linen regression smoke tests on `apply_lora_on_base_params` and `unapply_lora_from_base_params` (no existing unit test for these helpers in the tree)

**New unit tests** (`tests/unit/grpo_nnx_test.py`, 8 tests):
- 5 on `grpo_loss_fn_nnx`: `LossAux` shape parity, signature compatibility, identical-policy/reference → zero KL, `grpo_beta=0` → `aux.avg_kl=None`, finite policy grads
- 1 on `compute_log_probs_nnx`: shape `[B, S] → [B, S-1]`
- 2 Linen regression smoke tests on `grpo_loss_fn` and `compute_log_probs` (the existing Linen integration test is TPU-only and currently `@pytest.mark.skip`)

**Modified test:** `tests/unit/maxengine_test.py` swaps `test_lora_raises_for_nnx` (asserted `NotImplementedError`) for `test_lora_load_single_adapter_reaches_loader_on_nnx` (asserts `FileNotFoundError` from the loader).

**Existing Linen tests:** untouched and still pass; `pure_nnx=False` stays default.

**Test results:** 198 passed, 1 skipped (pre-existing CPU-only skip) across the broader NNX regression sweep — `maxengine_test`, `dpo_nnx_test`, `train_nnx_test`, `lora_utils_nnx_test`, `grpo_nnx_test`, `train_state_nnx_test`, `train_utils_nnx_test`, `gradient_accumulation_nnx_test`, `linen_nnx_converter_test`, `compare_linen_nnx_checkpoint_test`.

**Linting:** `bash lint.sh` — pyink + pylint 10.00/10.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/experimental/rl/grpo_trainer.py [hunk 1]
```diff
@@ -335,34 +339,190 @@ def grpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_
...
+      "learning/avg_reward_std": aux.avg_reward_std,
+      "learning/avg_advantage": aux.avg_advantage,
+      "learning/avg_kl": aux.avg_kl,
+      "learning/completion_length": aux.completion_length,
+      "learning/moe_lb_loss": aux.moe_lb_loss,
+      "learning/total_weights": aux.total_weights,
+      "learning/grad_norm": max_utils.l2norm_pytree(grads),
+      "learning/raw_grad_norm": max_utils.l2norm_pytree(raw_grads),
+  }
+  _, new_policy_params, _ = nnx.split(new_state.model, nnx.Param, ...)
```
Comment (cgarciae) [inline src/maxtext/experimental/rl/grpo_trainer.py]:
```suggestion
  new_policy_params = nnx.state(new_state.model, nnx.Param)
```

Comment (ecnal-cienet) [reply src/maxtext/experimental/rl/grpo_trainer.py]: Applied. Thank you.

File: src/maxtext/utils/lora_utils.py [hunk 2]
```diff
@@ -610,3 +645,186 @@ def _map_to_state(path, variable):
...
+        recurse(base_node[name], lora_child, f"{path}.{name}")
+      elif lora_child is not None:
+        if name not in ("lora_a.kernel", "lora_b.kernel"):
+          raise ValueError(f"Unexpected non-lora key ({path}.{name}) in lora_params")
+        lora_b = lora_node["lora_a.kernel"]
+        lora_a = lora_node["lora_b.kernel"]
+        base_node["kernel"] = lora_update_or_base(base_node["kernel"], lora_a, lora_b)
+        return
+
+  recurse(base_params, lora_params, "")
```
Comment (bvandermoon) [inline src/maxtext/utils/lora_utils.py:697]: Why recursion here?

**Review by bvandermoon — APPROVED** [review]

File: src/maxtext/utils/lora_utils.py [hunk 3]
```diff
@@ -610,3 +645,186 @@ def _map_to_state(path, variable):
...
+    for name, lora_child in lora_node.items():
+      if _is_nnx_branch(lora_child):
+        recurse(base_node[name], lora_child, f"{path}.{name}")
+      elif lora_child is not None:
+        if name not in ("lora_a.kernel", "lora_b.kernel"):
+          raise ValueError(f"Unexpected non-lora key ({path}.{name}) in lora_params")
+        lora_b = lora_node["lora_a.kernel"]
+        lora_a = lora_node["lora_b.kernel"]
+        base_node["kernel"] = lora_update_or_base(base_node["kernel"], lora_a, lora_b)
+        return
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/utils/lora_utils.py]: fixed

File: src/maxtext/utils/lora_utils.py [hunk 4]
```diff
@@ -610,3 +645,186 @@ def _map_to_state(path, variable):
...
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/utils/lora_utils.py]: fixed

File: src/maxtext/utils/lora_utils.py (hunk 2 above)
Comment (ecnal-cienet) [reply src/maxtext/utils/lora_utils.py:697]: The recursion mirrors the existing Linen sibling apply_lora_on_base_params (which has used the same `apply_lora_recursively` walker since it was written), so apply and unapply behave consistently on both paths.

Functionally, it's also the natural shape for this walk: each lora subtree pairs `lora_a.kernel` and `lora_b.kernel` at the same leaf-level node and we need both siblings together to update the single corresponding kernel in the base subtree. A `jax.tree_util.tree_map_with_path` rewrite would need a parent-grouping pass (or carrying both subtrees and reconstructing siblings from paths), so it'd be more code rather than less.

I refactored the body to make the recursion shape obvious — leaf detection is now split out as `if "lora_a.kernel" in lora_node: ... return` at the top of `recurse`, instead of detecting leaves mid-iteration with a `break/return`.

**Review by cgarciae — APPROVED** [review]
